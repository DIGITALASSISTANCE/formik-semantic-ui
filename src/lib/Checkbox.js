import React, {Component} from 'react';
import {Form, Checkbox} from 'semantic-ui-react';
import {Field} from 'formik';

let fieldCounter = 0;
class FormikCheckbox extends Component {
  constructor(props) {
    super(props);
    this.id = props.id || `field_checkbox_${fieldCounter++}`;
  }

  render() {
    const {name, label, inputProps = {}, fieldProps = {}} = this.props;
    const {onChange, ...safeInputProps} = inputProps;
    return (
      <Field
        name={name}
        render={({field, form}) => {
          const error = form.touched[name] && form.errors[name];
          return (
            <Form.Field error={!!error} {...fieldProps}>
              <Checkbox
                id={this.id}
                name={name}
                label={label}
                {...safeInputProps}
                checked={field.value}
                onChange={(e, {name, checked}) => {
                  form.setFieldValue(name, checked, true);
                  onChange && onChange(e, {name, value: checked});
                }}
              />
              {error && (
                <span className="sui-error-message">{form.errors[name]}</span>
              )}
            </Form.Field>
          );
        }}
      />
    );
  }
}

export default FormikCheckbox;
