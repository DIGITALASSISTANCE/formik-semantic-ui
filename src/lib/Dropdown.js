import React, { Component } from "react";
import { Form, Dropdown } from "semantic-ui-react";
import { Field } from "formik";

let fieldCounter = 0;
class FormikDropdown extends Component {
  constructor(props) {
    super(props);
    this.id = props.id || `field_dropdown_${fieldCounter++}`;
  }

  render() {
    const {
      name,
      label,
      options,
      inputProps = {},
      fieldProps = {}
    } = this.props;
    return (
      <Field
        name={name}
        render={({ field, form }) => {
          const error = form.touched[name] && form.errors[name];
          return (
            <Form.Field error={!!error} {...fieldProps}>
              {!!label && (
                <label htmlFor={this.id} onClick={() => this._dropdown.open()}>
                  {label}
                </label>
              )}
              <Dropdown
                ref={el => (this._dropdown = el)}
                id={this.id}
                name={name}
                options={options}
                selectOnBlur={false}
                selectOnNavigation={false}
                selection
                {...inputProps}
                value={field.value}
                onChange={(e, { name, value }) => {
                  form.setFieldValue(name, value, true);
                }}
              />
              {error && (
                <span className="sui-error-message">{form.errors[name]}</span>
              )}
            </Form.Field>
          );
        }}
      />
    );
  }
}

export default FormikDropdown;
