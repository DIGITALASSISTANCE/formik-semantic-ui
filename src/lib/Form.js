import React from "react";
import hoistNonReactStatics from "hoist-non-react-statics";
import { Formik } from "formik";
import { Form as SemanticForm } from "semantic-ui-react";

const FormikForm = ({
  className,
  inverted,
  size,
  ignoreLoading,
  children,
  render,
  component,
  ...formikProps
}) => {
  if (component) {
    console.error("The formik component prop is not supported in this wrapper");
    console.error(
      "Import the Formik control from the formik package to use this scenario"
    );
    throw new Error("Unsupported prop");
  }
  return (
    <Formik {...formikProps}>
      {renderProps => {
        const { handleSubmit, isSubmitting } = renderProps;
        return (
          <SemanticForm
            {...{ className, inverted, size }}
            onSubmit={handleSubmit}
            loading={isSubmitting && !ignoreLoading}
          >
            {typeof children === "function" || typeof render === "function"
              ? (render || children)(renderProps)
              : children}
          </SemanticForm>
        );
      }}
    </Formik>
  );
};

const Form = hoistNonReactStatics(FormikForm, SemanticForm);
Form.Children = React.Fragment;

export default Form;
