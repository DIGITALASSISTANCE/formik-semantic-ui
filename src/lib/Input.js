import React, { Component } from "react";
import { Form, Input } from "semantic-ui-react";
import { Field } from "formik";

let fieldCounter = 0;
class FormikInput extends Component {
  constructor(props) {
    super(props);
    this.id = props.id || `field_input_${fieldCounter++}`;
  }

  render() {
    const { name, label, inputProps = {}, fieldProps = {} } = this.props;
    return (
      <Field
        name={name}
        render={({ field, form }) => {
          const error = form.touched[name] && form.errors[name];
          return (
            <Form.Field error={!!error} {...fieldProps}>
              {!!label && <label htmlFor={this.id}>{label}</label>}
              <Input
                id={this.id}
                name={name}
                value={field.value}
                onChange={(e, { name, value }) => {
                  form.setFieldValue(name, value, true);
                }}
                {...inputProps}
              />
              {error && (
                <span className="sui-error-message">{form.errors[name]}</span>
              )}
            </Form.Field>
          );
        }}
      />
    );
  }
}

export default FormikInput;
