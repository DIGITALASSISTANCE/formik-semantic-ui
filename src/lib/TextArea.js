import React, {Component} from 'react';
import {Form, TextArea} from 'semantic-ui-react';
import {Field} from 'formik';

let fieldCounter = 0;
class FormikTextArea extends Component {
  constructor(props) {
    super(props);
    this.id = props.id || `field_textarea_${fieldCounter++}`;
  }

  render() {
    const {name, label, inputProps = {}, fieldProps = {}} = this.props;
    return (
      <Field
        name={name}
        render={({field, form}) => {
          const error = form.touched[name] && form.errors[name];
          return (
            <Form.Field error={!!error} {...fieldProps}>
              {!!label && <label htmlFor={this.id}>{label}</label>}
              <TextArea
                id={this.id}
                name={name}
                value={field.value}
                onChange={(e, {name, value}) => {
                  form.setFieldValue(name, value, true);
                }}
                rows={4}
                {...inputProps}
              />
              {error && (
                <span className="sui-error-message">{form.errors[name]}</span>
              )}
            </Form.Field>
          );
        }}
      />
    );
  }
}

export default FormikTextArea;
