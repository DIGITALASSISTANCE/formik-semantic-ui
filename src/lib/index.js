import "./index.css";

export { default as Input } from "./Input";
export { default as Checkbox } from "./Checkbox";
export { default as Dropdown } from "./Dropdown";
export { default as TextArea } from "./TextArea";

// export { default as DatePicker } from "./DatePicker";
// export { default as FileUpload } from "./FileUpload";
