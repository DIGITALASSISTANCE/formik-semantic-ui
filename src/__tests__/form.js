import React from 'react';
import {render, cleanup, fireEvent, wait} from 'react-testing-library';

import {Button, Form} from '../lib/index';

describe('Form', () => {
  afterEach(() => {
    cleanup();
  });

  it('child as a function', async () => {
    const handleSubmit = jest.fn();

    const {container, getByText} = render(
      <Form initialValues={{name: ''}} onSubmit={handleSubmit}>
        {({isSubmitting}) => (
          <React.Fragment>
            <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
          </React.Fragment>
        )}
      </Form>
    );

    const submitButton = getByText('Submit');

    await fireEvent.click(submitButton);
    await fireEvent.click(submitButton);

    await wait(() => {
      expect(handleSubmit).toHaveBeenCalled();
    });

    expect(handleSubmit.mock.calls.length).toEqual(1);
    expect(container).toMatchSnapshot();
  });

  it('render props', async () => {
    const {container, getByText} = render(
      <Form
        initialValues={{name: ''}}
        onSubmit={() => {}}
        render={({isSubmitting}) => (
          <React.Fragment>
            <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
          </React.Fragment>
        )}
      />
    );

    const submitButton = getByText('Submit');
    await fireEvent.click(submitButton);
    expect(container).toMatchSnapshot();
  });

  it('ignore loading', async () => {
    const {container, getByText} = render(
      <Form
        initialValues={{name: ''}}
        ignoreLoading
        onSubmit={() => {}}
        render={({isSubmitting}) => (
          <React.Fragment>
            <Button.Submit disabled={isSubmitting}>Submit</Button.Submit>
          </React.Fragment>
        )}
      />
    );

    const submitButton = getByText('Submit');
    await fireEvent.click(submitButton);
    expect(container).toMatchSnapshot();
  });
});
