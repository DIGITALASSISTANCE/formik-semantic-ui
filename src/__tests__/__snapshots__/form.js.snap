// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form child as a function 1`] = `
<div>
  <form
    class="ui loading form"
  >
    <button
      type="submit"
      class="ui primary disabled button"
      role="button"
      disabled=""
      tabindex="-1"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form ignore loading 1`] = `
<div>
  <form
    class="ui form"
  >
    <button
      type="submit"
      class="ui primary disabled button"
      role="button"
      disabled=""
      tabindex="-1"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Form render props 1`] = `
<div>
  <form
    class="ui loading form"
  >
    <button
      type="submit"
      class="ui primary disabled button"
      role="button"
      disabled=""
      tabindex="-1"
    >
      Submit
    </button>
  </form>
</div>
`;
